packages/frontend/src/pages/campaigns/index.tsx

```diff
@@ -77,25 +64,72 @@ export const Campaigns = () => {
     const { searchQuery, setSearchQuery } = useSearch();
     const { loading, kpiTokens } = useKPITokens(searchQuery);
 
-    //  show filters
-    const [filtersOpen, setFilterOpen] = useState(false);
+    //  toggle filters
     const toggleFilters = () => setFilterOpen(!filtersOpen);
+    const [filtersOpen, setFilterOpen] = useState(false);
 
     // page settings
     const [pageSize, setPageSize] = useState(12);
     const [ordering, setOrdering] = useState<SelectOption>(ORDERING_OPTIONS[0]);
+
+    // filters
     const [campaignState, setCampaignState] = useState<SelectOption>(
         STATE_OPTIONS[0]
     );
+    const [selectedTemplates, setSelectedTemplates] = useState<Set<number>>(
+        new Set<number>()
+    );
+
+    const [selectedOracles, setSelectedOracles] = useState<Set<number>>(
+        new Set<number>()
+    );
+
+    const handleSelectedTemplatesUpdate = useCallback(
+        (newSelectedTemplates: Set<number>) =>
+            setSelectedTemplates(new Set(newSelectedTemplates)),
+        []
+    );
+
+    const handleSelectedOraclesTemplatesUpdate = useCallback(
+        (newSelectedOracles: Set<number>) =>
+            setSelectedOracles(new Set(newSelectedOracles)),
+        []
+    );
 
     // filter results
-    const filteredTokens = useMemo(() => {
-        return filterKPITokens(
-            Object.values(kpiTokens),
+    const filteredKPITokensByState = useMemo(() => {
+        const tokens = Object.values(kpiTokens);
+
+        return filterStateKPITokens(
+            tokens,
             campaignState.value as unknown as CampaignState
         );
     }, [kpiTokens, campaignState]);
 
+    const hasOracleTemplate = useMemo(
+        () => (token: KPIToken) =>
+            oraclesTemplateIds(token).some((oracleId) => {
```

Comment:
Looks like the `oraclesTemplateIds` function below is only used here, so I think we can just write this as:

```
token.oracles.map((oracle) => oracle.template.id).some((oracleId) => {
    return selectedOracles.has(Number(oracleId));
})
```

Reply:
Also, I'm not sure why the `Number(...)` casts everywhere on the template id. It should already be of type number.

Reply:
I'll take a look on this. But compiler issues.

Reply:
`Number` is not needed in this case.

Reply:
fixed: oraclesTemplateIds